File: ts/src/batch.ts
/**
 * Batch: accumulates writes and removes, commits once on commit().
 */

import git from 'isomorphic-git';
import {
  MODE_BLOB,
  MODE_LINK,
  MODE_TREE,
  FileNotFoundError,
  IsADirectoryError,
  type FsModule,
} from './types.js';
import { normalizePath } from './paths.js';
import { modeFromDisk, walkTo, existsAtPath, type TreeWrite } from './tree.js';

import type { FS } from './fs.js';
import type { GitStore } from './gitstore.js';

/**
 * Accumulates writes and removes, then commits all changes atomically
 * when `commit()` is called.
 */
export class Batch {
  private _fs: FS;
  private _store: GitStore;
  private _fsModule: FsModule;
  private _gitdir: string;
  private _message: string | null;
  private _operation: string | null;
  private _writes = new Map<string, TreeWrite>();
  private _removes = new Set<string>();
  private _closed = false;

  /** The resulting FS snapshot after commit. Null until commit() completes. */
  result: FS | null = null;

  constructor(fs: FS, message?: string | null, operation?: string | null) {
    if (!fs._writable) {
      throw new Error('Cannot batch on a read-only snapshot');
    }
    this._fs = fs;
    this._store = fs._store;
    this._fsModule = fs._store._fsModule;
    this._gitdir = fs._store._gitdir;
    this._message = message ?? null;
    this._operation = operation ?? null;
  }

  private _checkOpen(): void {
    if (this._closed) throw new Error('Batch is closed');
  }

  /**
   * Stage a blob write. Creates the blob immediately in the object store.
   */
  async write(path: string, data: Uint8Array, opts?: { mode?: string }): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    this._removes.delete(normalized);
    const blobOid = await git.writeBlob({ fs: this._fsModule, gitdir: this._gitdir, blob: data });
    this._writes.set(normalized, {
      oid: blobOid,
      mode: opts?.mode ?? MODE_BLOB,
    });
  }

  /**
   * Stage a write from a local file. Reads the file and creates the blob.
   */
  async writeFromFile(
    path: string,
    localPath: string,
    opts?: { mode?: string },
  ): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    this._removes.delete(normalized);

    const detectedMode = await modeFromDisk(this._fsModule, localPath);
    const mode = opts?.mode ?? detectedMode;
    const data = (await this._fsModule.promises.readFile(localPath)) as Uint8Array;
    const blobOid = await git.writeBlob({ fs: this._fsModule, gitdir: this._gitdir, blob: data });
    this._writes.set(normalized, { oid: blobOid, mode });
  }

  /**
   * Stage a symlink write.
   */
  async writeSymlink(path: string, target: string): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    this._removes.delete(normalized);
    const data = new TextEncoder().encode(target);
    const blobOid = await git.writeBlob({ fs: this._fsModule, gitdir: this._gitdir, blob: data });
    this._writes.set(normalized, { oid: blobOid, mode: MODE_LINK });
  }

  /**
   * Stage a file removal.
   */
  async remove(path: string): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    const pendingWrite = this._writes.has(normalized);
    const existsInBase = await existsAtPath(
      this._fsModule,
      this._gitdir,
      this._fs._treeOid,
      normalized,
    );

    if (!pendingWrite && !existsInBase) {
      throw new FileNotFoundError(normalized);
    }

    // Don't allow removing directories
    if (existsInBase) {
      const entry = await walkTo(this._fsModule, this._gitdir, this._fs._treeOid, normalized);
      if (entry.mode === MODE_TREE) {
        throw new IsADirectoryError(normalized);
      }
    }

    this._writes.delete(normalized);
    if (existsInBase) {
      this._removes.add(normalized);
    }
  }

  /**
   * Commit all accumulated changes. Returns the new FS snapshot.
   */
  async commit(): Promise<FS> {
    if (this._closed) throw new Error('Batch is already committed');
    this._closed = true;

    if (this._writes.size === 0 && this._removes.size === 0) {
      this.result = this._fs;
      return this._fs;
    }

    this.result = await this._fs._commitChanges(
      this._writes,
      this._removes,
      this._message,
      this._operation,
    );
    return this.result;
  }
}

File: rs/src/batch.rs
use std::path::Path;

use crate::error::{Error, Result};
use crate::fs::{Fs, TreeWrite};
use crate::tree;
use crate::types::{MODE_BLOB, MODE_LINK};

/// Accumulates writes and commits them atomically when consumed.
///
/// `commit(self)` takes ownership, so the compiler enforces that
/// no further writes happen after committing.
pub struct Batch {
    pub(crate) fs: Fs,
    pub(crate) writes: Vec<(String, Option<TreeWrite>)>,
    pub(crate) removes: Vec<String>,
    pub(crate) message: Option<String>,
    pub(crate) operation: Option<String>,
    pub(crate) closed: bool,
}

impl Batch {
    fn require_open(&self) -> Result<()> {
        if self.closed {
            Err(Error::BatchClosed)
        } else {
            Ok(())
        }
    }

    /// Write raw bytes to `path`.
    pub fn write(&mut self, path: &str, data: &[u8]) -> Result<()> {
        self.write_with_mode(path, data, MODE_BLOB)
    }

    /// Write raw bytes to `path` with an explicit file mode.
    pub fn write_with_mode(&mut self, path: &str, data: &[u8], mode: u32) -> Result<()> {
        self.require_open()?;
        let path = crate::paths::normalize_path(path)?;

        let tw = self.fs.with_repo(|repo| {
            let blob_oid = repo.write_blob(data).map_err(Error::git)?;
            Ok(TreeWrite {
                data: data.to_vec(),
                oid: blob_oid.detach(),
                mode,
            })
        })?;

        // Remove from removes if present
        self.removes.retain(|p| p != &path);
        // Remove existing write for same path
        self.writes.retain(|(p, _)| p != &path);
        self.writes.push((path, Some(tw)));
        Ok(())
    }

    /// Write the contents of a disk file to `path`.
    pub fn write_from_file(&mut self, path: &str, src: &Path) -> Result<()> {
        self.require_open()?;
        let data = std::fs::read(src).map_err(|e| Error::io(src, e))?;
        let mode = tree::mode_from_disk(src).unwrap_or(MODE_BLOB);
        self.write_with_mode(path, &data, mode)
    }

    /// Write a symlink at `path`.
    pub fn write_symlink(&mut self, path: &str, target: &str) -> Result<()> {
        self.require_open()?;
        let path = crate::paths::normalize_path(path)?;

        let tw = self.fs.with_repo(|repo| {
            let blob_oid = repo.write_blob(target.as_bytes()).map_err(Error::git)?;
            Ok(TreeWrite {
                data: target.as_bytes().to_vec(),
                oid: blob_oid.detach(),
                mode: MODE_LINK,
            })
        })?;

        self.removes.retain(|p| p != &path);
        self.writes.retain(|(p, _)| p != &path);
        self.writes.push((path, Some(tw)));
        Ok(())
    }

    /// Mark `path` for removal.
    pub fn remove(&mut self, path: &str) -> Result<()> {
        self.require_open()?;
        let path = crate::paths::normalize_path(path)?;

        // Remove from writes if present
        self.writes.retain(|(p, _)| p != &path);

        if !self.removes.contains(&path) {
            self.removes.push(path);
        }
        Ok(())
    }

    /// Commit all accumulated writes. Consumes the `Batch` and returns the new `Fs` snapshot.
    pub fn commit(mut self) -> Result<Fs> {
        self.closed = true;

        if self.writes.is_empty() && self.removes.is_empty() {
            return Ok(self.fs);
        }

        // Merge removes into writes as (path, None)
        let mut all_writes = self.writes;
        for path in self.removes {
            all_writes.push((path, None));
        }

        let message = self.message.unwrap_or_else(|| {
            crate::paths::format_commit_message(
                self.operation.as_deref().unwrap_or("batch"),
                None,
            )
        });

        self.fs.commit_changes(&all_writes, &message)
    }

    /// Returns `true` if `commit()` has already been called.
    pub fn is_closed(&self) -> bool {
        self.closed
    }
}
